overlord/devicestate/devicemgr.go

```diff
@@ -679,6 +679,35 @@ func (m *DeviceManager) ensureCloudInitRestricted() error {
 			statusMsg = "failed to transition to done or error state after 5 minutes"
 		}
 
+		// we should always have a model if we are seeded and are not on classic
+		model, err := m.Model()
+		if err != nil {
+			return err
+		}
+
+		// For UC20, we want to always disable cloud-init after it has run on
+		// first boot unless we are in a "real cloud", i.e. not using NoCloud,
+		// or if we installed cloud-init configuration from the gadget
+		if model.Grade() != asserts.ModelGradeUnset {
+			// look for cloud.conf in the installed gadget snap
+			deviceCtx, err := DeviceCtx(m.state, nil, nil)
+			if err != nil {
+				return err
+			}
+
+			gadgetInfo, err := snapstate.GadgetInfo(m.state, deviceCtx)
+			if err != nil {
+				return err
+			}
+
+			if !sysconfig.HasGadgetCloudConf(gadgetInfo.MountDir()) {
+				// no gadget cloud conf, so disable cloud-init if it's NoCloud
+				// (it will be restricted for any other cloud to that specific
+				// cloud provider datasource)
+				opts.DisableNoCloud = true
```

Comment:
mmh, I'm thinking again about this, is there are strong reason to let NoCloud run more than for first boot even if the gadget had a cloud.conf ? basically, I'm thinking to simplify and

UC20 => DisableNoCloud = true

Reply:
done, as discussed on IRC, it's fine to disable NoCloud even if there's a gadget using NoCloud since we don't want NoCloud to be used across multiple boots since it could be insecure, etc.